crypto/dilithium/p_dilithium_test.cc

```diff
@@ -0,0 +1,152 @@
+  ASSERT_NE(dilithium_pkey2, nullptr);
+
+  EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx2));
+  EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx2, &dilithium_pkey2));
+  ASSERT_NE(dilithium_pkey2->pkey.ptr, nullptr);
+
+  EXPECT_EQ(0, EVP_PKEY_cmp(dilithium_pkey1, dilithium_pkey2));
+  EXPECT_EQ(1, EVP_PKEY_cmp(dilithium_pkey1, dilithium_pkey1));
+  EXPECT_EQ(1, EVP_PKEY_cmp(dilithium_pkey2, dilithium_pkey2));
+
+  EVP_PKEY_CTX_free(dilithium_pkey_ctx1);
+  EVP_PKEY_CTX_free(dilithium_pkey_ctx2);
+}
+
+TEST(Dilithium3Test, NewKeyFromBytes) {
+  // Source key
+  EVP_PKEY_CTX *dilithium_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  ASSERT_NE(dilithium_pkey_ctx, nullptr);
+
+  EVP_PKEY *dilithium_pkey = EVP_PKEY_new();
+  ASSERT_NE(dilithium_pkey, nullptr);
+
+  EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx));
+  EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx, &dilithium_pkey));
+  ASSERT_NE(dilithium_pkey->pkey.ptr, nullptr);
+  const DILITHIUM3_KEY *dilithium3Key = (DILITHIUM3_KEY *)(dilithium_pkey->pkey.ptr);
+
+  // New raw public key
+  EVP_PKEY *new_public = EVP_PKEY_new_raw_public_key(EVP_PKEY_DILITHIUM3,
+                                                     NULL,
+                                                     dilithium3Key->pub,
+                                                     DILITHIUM3_PUBLIC_KEY_BYTES);
+  ASSERT_NE(new_public, nullptr);
+
+  uint8_t *buf = nullptr;
+  size_t buf_size;
+  EXPECT_FALSE(EVP_PKEY_get_raw_private_key(new_public, buf, &buf_size));
+  uint32_t err = ERR_get_error();
+  EXPECT_EQ(ERR_LIB_EVP, ERR_GET_LIB(err));
+  EXPECT_EQ(EVP_R_NOT_A_PRIVATE_KEY, ERR_GET_REASON(err));
+
+  // EVP_PKEY_cmp just compares the public keys so this should return 1
+  EXPECT_EQ(1, EVP_PKEY_cmp(dilithium_pkey, new_public));
+
+  // New raw private key
+  EVP_PKEY *new_private = EVP_PKEY_new_raw_private_key(EVP_PKEY_DILITHIUM3,
+                                                       NULL,
+                                                       dilithium3Key->priv,
+                                                       DILITHIUM3_PRIVATE_KEY_BYTES);
+  ASSERT_NE(new_private, nullptr);
+  const DILITHIUM3_KEY *newDilithium3Key = (DILITHIUM3_KEY *)(new_private->pkey.ptr);
+  EXPECT_EQ(0, OPENSSL_memcmp(dilithium3Key->priv, newDilithium3Key->priv,
+                              DILITHIUM3_PRIVATE_KEY_BYTES));
+
+  EVP_PKEY_CTX_free(dilithium_pkey_ctx);
+  EVP_PKEY_free(new_public);
+  EVP_PKEY_free(new_private);
+}
+
+TEST(Dilithium3Test, KeySize) {
+  EVP_PKEY_CTX *dilithium_pkey_ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_DILITHIUM3, nullptr);
+  ASSERT_NE(dilithium_pkey_ctx, nullptr);
+
+  EVP_PKEY *dilithium_pkey = EVP_PKEY_new();
+  ASSERT_NE(dilithium_pkey, nullptr);
+
+  EXPECT_TRUE(EVP_PKEY_keygen_init(dilithium_pkey_ctx));
+  EXPECT_TRUE(EVP_PKEY_keygen(dilithium_pkey_ctx, &dilithium_pkey));
+
+  EXPECT_EQ(DILITHIUM3_PUBLIC_KEY_BYTES + DILITHIUM3_PRIVATE_KEY_BYTES, EVP_PKEY_size(dilithium_pkey));
```

Comment:
doesn't `EVP_PKEY_size` return the signature size?

Reply:
Thanks, fixed in https://github.com/jakemas/aws-lc/commit/945faef337b75019d1afbb782864312397ce8ca6